datastore.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func (d *Datastore) Put(ctx context.Context, key ds.Key, value []byte) error {
+	txn, _ := d.NewTransaction(ctx, false)
+	err := txn.Put(ctx, key, value)
+	if err != nil {
+		return err
+	}
+	return txn.Commit(ctx)
+}
+
+func (d *Datastore) Delete(ctx context.Context, key ds.Key) error {
+	txn, _ := d.NewTransaction(ctx, false)
+	err := txn.Delete(ctx, key)
+	if err != nil {
+		return err
+	}
+	return txn.Commit(ctx)
+}
+
+func (d *Datastore) Sync(ctx context.Context, _ ds.Key) error {
+	return d.db.Backend().Flush()
+}
+
+func (d *Datastore) Close() error {
+	return d.db.Close()
+}
+
+func (d *Datastore) Batch(ctx context.Context) (ds.Batch, error) {
+	return d.NewTransaction(ctx, false)
+}
+
+func (d *Datastore) NewTransaction(ctx context.Context, readOnly bool) (ds.Txn, error) {
+	batch := d.db.Batch()
+	return &txn{
+		batch:    batch,
+		table:    d.table,
+		readOnly: readOnly,
+	}, nil
+}
+
+type txn struct {
+	table    bond.Table[*DatastoreEntry]
+	batch    bond.Batch
+	readOnly bool
+}
+
+var _ ds.Txn = &txn{}
+
+func (t *txn) Put(ctx context.Context, key ds.Key, value []byte) error {
+	if t.readOnly {
+		return fmt.Errorf("readonly txn")
+	}
+	return t.table.Upsert(ctx, []*DatastoreEntry{{Key: key.String(), Value: value}}, func(old, new *DatastoreEntry) *DatastoreEntry {
+		return new
+	}, t.batch)
+}
+
+func (t *txn) Delete(ctx context.Context, key ds.Key) error {
+	return t.table.Delete(ctx, []*DatastoreEntry{{Key: key.String()}}, t.batch)
+}
+
+func (t *txn) Get(ctx context.Context, key ds.Key) ([]byte, error) {
+	entries, err := t.table.Get(ctx, bond.NewSelectorPoint[*DatastoreEntry](&DatastoreEntry{Key: key.String()}), t.batch)
+	if err == nil {
+		return entries[0].Value, nil
+	}
+	if strings.Contains(err.Error(), "not found") {
+		return nil, ds.ErrNotFound
+	}
+	return nil, err
+}
+
+func (t *txn) Query(ctx context.Context, userQuery query.Query) (query.Results, error) {
+	prefix := ds.NewKey(userQuery.Prefix).String()
+	if prefix != "/" {
+		prefix = prefix + "/"
+	}
+	reverse := false
+
+	if len(userQuery.Orders) > 0 {
+		switch userQuery.Orders[0].(type) {
+		case query.OrderByKey, *query.OrderByKey:
+		// We order by key by default.
+		case query.OrderByKeyDescending, *query.OrderByKeyDescending:
+			// Reverse order by key
+			reverse = true
+		default:
+			// Skip the stuff we can't apply.
+			baseQuery := userQuery
+			baseQuery.Limit = 0
+			baseQuery.Offset = 0
+			baseQuery.Orders = nil
+
+			// perform the base query.
+			res, err := t.Query(ctx, baseQuery)
+			if err != nil {
+				return nil, err
+			}
+
+			// fix the query
+			res = query.ResultsReplaceQuery(res, userQuery)
+
+			// Remove the parts we've already applied.
+			naiveQuery := userQuery
+			naiveQuery.Prefix = ""
+			naiveQuery.Filters = nil
+
+			// Apply the rest of the query
+			return query.NaiveQueryApply(naiveQuery, res), nil
+		}
+	}
+
+	qrb := query.NewResultBuilder(userQuery)
+
+	qrb.Process.Go(func(proc goprocess.Process) {
```

Comment:
whats the purpose of this...? spinning up a new go routine for a query seems strange..?

Reply:
yes, query happens in an another go routine and we return a channel immediately. Later results are pushed into channel one by one

https://github.com/ipfs/go-datastore/blob/v0.6.0/query/query.go#L127

Reply:
lets do it all sync and leave to application layer.

if there are method as a result we don't support, implement the method and return an errors.New("dsbond: ___ method is not supported")